core/rest-client/src/http_client.rs

```diff
@@ -0,0 +1,425 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use crate::sgx_reexport_prelude::*;
+
+use crate::{error::Error, Query, RestPath};
+use codec::Encode;
+use http::{
+	header::{HeaderName, AUTHORIZATION, CONTENT_LENGTH, CONTENT_TYPE, USER_AGENT},
+	HeaderValue,
+};
+use http_req::{
+	request::{Method, Request},
+	response::{Headers, Response},
+	uri::Uri,
+};
+use log::*;
+use std::{
+	collections::HashMap,
+	str::FromStr,
+	string::{String, ToString},
+	time::Duration,
+	vec::Vec,
+};
+use url::Url;
+
+pub type EncodedBody = Vec<u8>;
+
+/// Simple trait to send HTTP request, based on `http_req`.
+///
+/// Automatically upgrades to TLS in case the base URL contains 'https'
+pub trait SendHttpRequest {
+	fn send_request<U, T>(
+		&self,
+		base_url: Url,
+		method: Method,
+		params: U,
+		query: Option<&Query<'_>>,
+		maybe_body: Option<String>,
+	) -> Result<(Response, EncodedBody), Error>
+	where
+		T: RestPath<U>;
+}
+
+/// HTTP client implementation
+///
+/// wrapper for the `http_req` library that adds the necessary headers and body to a request
+pub struct HttpClient {
+	send_null_body: bool,
+	timeout: Option<Duration>,
+	headers: Headers,
+	authorization: Option<String>,
+}
+
+impl Default for HttpClient {
+	fn default() -> Self {
+		HttpClient {
+			send_null_body: true,
+			timeout: None,
+			headers: Headers::new(),
+			authorization: None,
+		}
+	}
+}
+
+impl HttpClient {
+	pub fn new(
+		send_null_body: bool,
+		timeout: Option<Duration>,
+		headers: Option<Headers>,
+		authorization: Option<String>,
+	) -> Self {
+		HttpClient {
+			send_null_body,
+			timeout,
+			headers: headers.unwrap_or_else(Headers::new),
+			authorization,
+		}
+	}
+
+	/// Set credentials for HTTP Basic authentication.
+	pub fn set_auth(&mut self, user: &str, pass: &str) {
```

Comment:
Have you thought about this pattern to allow concatenation of configuration calls?

`pub fn with_auth(self, user, pass) -> Self`

Reply:
That's true, returning some form of `self` will allow the caller to handle this better. If I want to return `Self` though, I have to change the signature to `mut self`. 
This is another instance of just copying from restson and never using it 😄

Reply:
After giving it some thought, I'm not sure what the best pattern is here. Returning `Self` will require a move when calling this method (and have CLion complain about style). In addition, the method below `set_header` would have to be changed accordingly, and it returns a result, which would then mean we return a moved `Self` in the form of `Result<Self, Error>` ?

Reply:
Yes, I am also not sure, but I was curious about your thoughts. 😄

It's totally fine with me how it is. I have just been having this question quite some time and I could not really figure out what is best.

So let's keep it. 👍